internal/transactionrequest.go

```diff
@@ -208,6 +229,47 @@ func (t *TxnReq) ProcessTransaction(ctx context.Context) (*fclient.RespSend, *ut
 	return &fclient.RespSend{PDUs: results}, nil
 }
 
+// updateUnsignedIfNeeded sets unsigned.prev_content to the previous membership event. This is
+// to avoid checking the mxid_mapping on join -> join (e.g. displayname changes) events, which
+// wouldn't be present.
+func (t *TxnReq) updateUnsignedIfNeeded(ctx context.Context, event gomatrixserverlib.PDU) (updated bool, err error) {
+	var membership string
+	membership, err = event.Membership()
+	if err != nil {
+		return false, err
+	}
+	var validRoomID *spec.RoomID
+	validRoomID, err = spec.NewRoomID(event.RoomID())
+	if err != nil {
+		return false, err
+	}
+
+	// get the current membership event
+	var currentStateEv gomatrixserverlib.PDU
+	currentStateEv, err = t.rsAPI.CurrentStateEvent(ctx, *validRoomID, event.Type(), string(event.SenderID()))
+	if err != nil {
+		return false, err
+	}
+	if currentStateEv != nil {
+		var currentMembership string
+		currentMembership, err = currentStateEv.Membership()
+		if err != nil {
+			return false, err
+		}
+
+		// if the currentMembership is join and the new membership as well (e.g. displayname change),
+		// update unsigned so VerifyEventSignatures does not try to validate the mxid_mapping
+		if currentMembership == spec.Join && currentMembership == membership {
+			err = event.SetUnsignedField("prev_content", currentStateEv.Content())
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		}
```

Comment:
Setting the unsigned field like this feels super hacky. Could we not just tell GMSL not to do the mxid mapping check? Probably not in a generic way that still makes sense with non-pseudoIDs huh?
We could pass `current_membership` to `VerifyEventSignatures` and do the `join == join` check in there, but then that would add another DB hit everywhere this is used. (only ~5-10 places it seems, but still not ideal)
Setting the unsigned field seems fragile if for whatever reason membership events start using this field in the future.

Reply:
Indeed, this is super hacky. (Looking at where we verify event signatures, this needs to be in other places as well)
I'd like to avoid having to add yet another DB hit for every call to `VerifyEventSignatures`, given it is used in the hotpath when receiving and storing events.
Adding `currentMembership` as a parameter just for it to be used for pseudoIDs doesn't feel right either.

Need to think about what's the best way to avoid this hack.